crates/curp/src/server/raw_curp/log.rs

```diff
@@ -23,13 +23,77 @@ use crate::{
     LogIndex,
 };
 
+/// Enum representing a range of values in the log.
+#[derive(Debug, Eq, PartialEq)]
+enum LogRange<T> {
+    /// Represents a range using the `Range` type.
+    Range(Range<T>),
+    /// Represents a range using the `RangeInclusive` type.
+    RangeInclusive(RangeInclusive<T>),
+}
+
+impl<T> RangeBounds<T> for LogRange<T> {
+    fn start_bound(&self) -> Bound<&T> {
+        match *self {
+            Self::Range(ref range) => range.start_bound(),
+            Self::RangeInclusive(ref range) => range.start_bound(),
+        }
+    }
+
+    fn end_bound(&self) -> Bound<&T> {
+        match *self {
+            Self::Range(ref range) => range.end_bound(),
+            Self::RangeInclusive(ref range) => range.end_bound(),
+        }
+    }
+}
+
+impl<T> From<Range<T>> for LogRange<T> {
+    fn from(range: Range<T>) -> Self {
+        Self::Range(range)
+    }
+}
+
+impl<T> From<RangeInclusive<T>> for LogRange<T> {
+    fn from(range: RangeInclusive<T>) -> Self {
+        Self::RangeInclusive(range)
+    }
+}
+
 /// Curp logs
 /// There exists a fake log entry 0 whose term equals 0
 /// For the leader, there should never be a gap between snapshot and entries
+///
+/// Examples:
+/// This example will describe the relationship among `batch_end`, `batch_limit`, `first_idx_in_cur_batch` and `cur_batch_size`
+/// if `batch_limit` = 11 and entry sizes is vec![1,5,6,2,3,4,2], then the relationship between entry size and `batch_end` looks like:
+/// -------------------------------------------
+/// `entry_size[i]`| 1 | 5 | 6 | 2 | 3 | 4 | 2 |
+/// ---------------+---------------------------+
+/// i              | 0 | 1 | 2 | 3 | 4 | 5 | 6 |
+/// ---------------+---+---+---+---+---+---+---+
+/// `batch_end[i]` | 1 | 2 | 4 | 5 | 0 | 0 | 0 |
+/// -------------------------------------------
+///                                  ↑
+///                       `first_idx_in_cur_batch`
+/// (0, `batch_end[0]`) = (0,1), which means the `entries[0..=1]` is a valid batch whose size is 1+5=6, less than the `batch_limit`
+/// (1, `batch_end[1]`) = (1,2), which means the `entries[1..=2]` is a valid batch whose size is 5+6=11, equal to the `batch_limit`
```

Comment:
I think making the batch_end exclusive could simplified the code even further, for instance batch_end[0] = 1 means entries[0..1]. Please refer to https://github.com/xline-kv/Xline/pull/704#issuecomment-2034304554